ipykernel/kernelbase.py

```diff
@@ -681,9 +672,17 @@ def shutdown_request(self, stream, ident, parent):
         )
 
         self._at_shutdown()
-        # call sys.exit after a short delay
-        loop = ioloop.IOLoop.current()
-        loop.add_timeout(time.time()+0.1, loop.stop)
+
+        # Flush to ensure reply is sent before stopping loop
+        self.control_stream.flush(zmq.POLLOUT)
+
+        self.log.debug('Stopping control ioloop')
+        control_io_loop = self.control_stream.io_loop
+        control_io_loop.add_callback(control_io_loop.stop)
+
+        self.log.debug('Stopping shell ioloop')
+        shell_io_loop = self.shell_stream.io_loop
+        shell_io_loop.add_callback(shell_io_loop.stop)
```

Comment:
I have been instrumenting the code to investigate the random failure, and it seems that `shell_io_loop.add_callback(shell_io_loop.stop)` does not always stop the shell ioloop for some reason.

Note: this function is called from the control thread, but `add_callback` is supposedly thread safe.